### PR TITLE
Deprecates absent types and removes associated planner logic

### PR DESCRIPTION
**Pull request by johnedquinn**
## Description
- As discussed with the rest of the PartiQL maintainers, the null and missing values have leaked into our type semantics -- causing issues when it comes to function resolution, unsafe casts, and dynamic functions.
- This PR deprecates NullType and MissingType and removes their usage in the planner. The implication is that ALL types may contain the null and missing values. See the CHANGELOG for all deprecated APIs.
- In this PR, literal NULL and MISSING values are typed as ANY (dynamic). In the future, we may want to pursue the "unknown" type that PostgreSQL internally uses.
- Smaller changes include:
  - I deprecated `AnyOfType`'s primary constructor. We've been plagued for too long by forgetting to use `.flatten()` .
  - The majority of this diff is me getting rid of "null" and "missing" in the testFixtures.

## Other Information
- Updated Unreleased Section in CHANGELOG: **YES**
- Any backward-incompatible changes? **YES**
  - Not an API breaking change, but a behavioral one.
- Any new external dependencies? **NO**
- Do your changes comply with the [Contributing Guidelines](https://github.com/partiql/partiql-lang-kotlin/blob/main/CONTRIBUTING.md)
  and [Code Style Guidelines](https://github.com/partiql/partiql-lang-kotlin/blob/main/CODE_STYLE.md)? **YES**

## License Information

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: partiql-planner/src/test/kotlin/org/partiql/planner/internal/typer/PlanTyperTestsPorted.kt [hunk 1]
```diff
@@ -2144,15 +2136,13 @@ class PlanTyperTestsPorted {
                     ) FROM <<
                         { 'a': { 'b': 1 } },
                         { 'a': { 'b': 'hello' } },
-                        { 'a': NULL },
```
Comment (johnedquinn) [inline partiql-planner/src/test/kotlin/org/partiql/planner/internal/typer/PlanTyperTestsPorted.kt:2147]: I updated this only because the literal NULL's type is unknown. Which leads to the inferred type to be ANY when used as an argument to a literal struct.

File: partiql-planner/src/test/kotlin/org/partiql/planner/internal/typer/PlanTyperTestsPorted.kt [hunk 2]
```diff
@@ -3127,55 +3104,8 @@ class PlanTyperTestsPorted {
...
-                        contentClosed = true,
-                        constraints = setOf(
-                            TupleConstraint.Open(false),
-                            TupleConstraint.UniqueAttrs(true),
-                            TupleConstraint.Ordered
-                        )
-                    )
-                )
-            ),
-            SuccessTestCase(
```
Comment (johnedquinn) [inline partiql-planner/src/test/kotlin/org/partiql/planner/internal/typer/PlanTyperTestsPorted.kt:3142]: This is deleted now since ALL types are nullable.

File: partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/FnResolver.kt [hunk 3]
```diff
@@ -440,8 +407,10 @@ internal class FnResolver(private val header: Header) {
         // This is not explicitly defined in the PartiQL Specification!!
         // This does not imply the ability to CAST; this defines function resolution behavior.
         private val precedence: Map<PartiQLValueType, Int> = listOf(
-            NULL,
-            MISSING,
+            @Suppress("DEPRECATION")
+            PartiQLValueType.NULL, // TODO: Remove
+            @Suppress("DEPRECATION")
+            PartiQLValueType.MISSING, // TODO: Remove
```
Comment (alancai98) [inline partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/FnResolver.kt:413]: should be removed per the TODO?

File: partiql-types/src/main/kotlin/org/partiql/types/StaticType.kt [hunk 4]
```diff
@@ -20,7 +20,7 @@ public sealed class StaticType {
          */
         @JvmStatic
         public fun unionOf(vararg types: StaticType, metas: Map<String, Any> = mapOf()): StaticType =
-            unionOf(types.toSet(), metas)
+            unionOf(types.toSet(), metas).flatten()
```
[the remark this answers is not in the document]
Comment (alancai98) [reply partiql-types/src/main/kotlin/org/partiql/types/StaticType.kt:23]: Side-note: there are probably a bunch of other places we now call `.flatten()` redundantly (especially in `PlanTyper`), perhaps we could fix or cut an issue to track fixing.

File: partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt [hunk 5]
```diff
@@ -515,20 +511,31 @@ internal class PlanTyper(
...
-            }.toSet()
-            val finalType = unionOf(pathTypes + toAddTypes).flatten()
-            return rex(finalType.swallowAny(), rexOpPathKey(root, key))
+            }
+            if (elementType.isEmpty()) {
+                handleAlwaysMissing()
+                return rex(ANY, rexOpPathKey(root, key))
+            }
+            // TODO: SwallowAny should happen by default
+            return rex(unionOf(elementType).swallowAny(), rexOpPathKey(root, key))
```
Comment (alancai98) [inline partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt:520]: perhaps we could get rid of `.swallowAny()` function overall? I think this PR https://github.com/partiql/partiql-lang-kotlin/pull/1395 fixes the flatten call when one of the types is `AnyType`.

File: partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt [hunk 6]
```diff
@@ -537,11 +544,21 @@ internal class PlanTyper(
...
-            val type = unionOf(paths.map { it.type }.toSet()).flatten()
+            // Determine output type
+            val type = when (paths.size) {
+                // Escape early since no inference could be made
+                0 -> {
+                    handleAlwaysMissing()
+                    return rex(ANY, Rex.Op.Path.Symbol(root, node.key))
+                }
+                // TODO: Flatten() should occur by default
+                else -> unionOf(paths.map { it.type }.toSet()).flatten()
```
Comment (alancai98) [inline partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt:555]: yea with the changes in `StaticType.unionOf()`, `flatten()` can be removed here

File: partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt [hunk 7]
```diff
@@ -562,15 +579,6 @@ internal class PlanTyper(
 
         private fun rexString(str: String) = rex(STRING, rexOpLit(stringValue(str)))
 
-        override fun visitRexOpPath(node: Rex.Op.Path, ctx: StaticType?): Rex {
-            val path = super.visitRexOpPath(node, ctx) as Rex
-            if (path.type == MISSING) {
```
Comment (alancai98) [inline partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt:567]: should this override be deleted? Should we check if the `path.type == ANY`?

File: partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt [hunk 8]
```diff
@@ -621,71 +625,29 @@ internal class PlanTyper(
...
-            // 7.1 All functions return MISSING when one of their inputs is MISSING (except `=`)
-            newArgs.forEach {
-                if (it.type == MissingType && !isNotMissable) {
-                    handleAlwaysMissing()
-                    return rex(MISSING, rexOpCallStatic(newFn, newArgs))
-                }
+            // Check literal missing inputs
+            val argAlwaysMissing = args.any {
+                val op = it.op as? Rex.Op.Lit ?: return@any false
+                op.value.type == PartiQLValueType.MISSING
```
Comment (alancai98) [inline partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt:636]: similar to other comment, should check if something is a missing literal by checking if it's the literal missing value?

```suggestion
                op.value == missingValue()
```

File: partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt [hunk 9]
```diff
@@ -1225,26 +1163,30 @@ internal class PlanTyper(
                 isClosed && isOrdered -> {
                     struct.fields.firstOrNull { entry -> binding.isEquivalentTo(entry.key) }?.let {
                         (sensitive(it.key) to it.value)
-                    } ?: (key to MISSING)
+                    } ?: run {
+                        return null
+                    }
```
Comment (alancai98) [inline partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt:1168]: think you don't need this `run`?

```suggestion
                    } ?: return null
```

File: partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt [hunk 10]
```diff
@@ -621,71 +625,29 @@ internal class PlanTyper(
...
-                        if (arg.type.isNullable()) {
-                            isNullable = true
-                            break
-                        }
-                    }
+            if (argAlwaysMissing) {
+                // TODO: The V1 branch has support for isMissable and isMissingCall. This codebase, however, does not
+                //  have support for these concepts yet. This specific commit (see Git blame) does not seek to add this
+                //  functionality. Below is a work-around for the lack of "isMissable" and "isMissingCall"
+                if (match.signature.name !in listOf("is_null", "is_missing", "eq")) {
```
Comment (alancai98) [inline partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt:642]: Do we also need to include the logical ops here (and, or, not)? When `AND` and `OR` are provided with a `NULL`/`MISSING` argument, they may still return a boolean:

```
PartiQL> NULL AND false;
==='
false
---
OK!
PartiQL> NULL OR true;
==='
true
---
OK!
```

File: partiql-planner/src/test/kotlin/org/partiql/planner/internal/typer/PlanTyperTestsPorted.kt [hunk 11]
```diff
@@ -635,26 +637,23 @@ class PlanTyperTestsPorted {
             SuccessTestCase(
                 name = "BITWISE_AND_NULL_OPERAND",
                 query = "1 & NULL",
-                expected = StaticType.NULL,
+                expected = unionOf(INT4, INT8, INT),
```
Comment (alancai98) [inline partiql-planner/src/test/kotlin/org/partiql/planner/internal/typer/PlanTyperTestsPorted.kt:640]: Hmm, curious why this will now give a union of these int types? Similarly for the below case w/ `MISSING`.

File: partiql-planner/src/test/kotlin/org/partiql/planner/internal/typer/PlanTyperTestsPorted.kt [hunk 12]
```diff
@@ -2506,37 +2494,27 @@ class PlanTyperTestsPorted {
...
-            ),
-            SuccessTestCase(
-                key = PartiQLTest.Key("basics", "case-when-22"),
-                catalog = "pql",
-                expected = unionOf(StaticType.INT4, StaticType.NULL, StaticType.MISSING),
-            ),
-            SuccessTestCase(
-                key = PartiQLTest.Key("basics", "case-when-23"),
-                catalog = "pql",
-                expected = StaticType.INT4,
```
Comment (alancai98) [inline partiql-planner/src/test/kotlin/org/partiql/planner/internal/typer/PlanTyperTestsPorted.kt:2539]: Any reason these tests are deleted?

**Review by alancai98 — COMMENTED** [review]
nice work in removing the absent types from the static type modeling. I think it helps simplify a lot of the typing and test fixture involving typing logic quite a bit. comments were mostly minor though I did have some questions about some behavioral changes in tests.

File: partiql-planner/src/testFixtures/resources/catalogs/default/pql/t_item.ion [hunk 13]
```diff
@@ -10,7 +10,7 @@
     },
     {
       name: "t_bool_nul",
-      type: ["bool","null"],
+      type: "bool",
```
[the remark this answers is not in the document]
Comment (johnedquinn) [reply partiql-planner/src/testFixtures/resources/catalogs/default/pql/t_item.ion:13]: Yeah, I thought to do that, but I opted to avoid doing that due to the diff on tests. I can either include that change or follow up with a clean up of the duplicate tests.

File: partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt (hunk 8 above)
Comment (johnedquinn) [reply partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt:636]: The suggestion is a method invocation resulting in a whole new Java class instance and uses the equals method to check the Java class and properties. The original is a check on an enum (integer). Also, this is similar to the reasoning behind PQLValue -- we should rely on our own type constructs rather than Java's.

File: partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/DynamicTyper.kt [hunk 14]
```diff
@@ -57,59 +54,63 @@ internal class DynamicTyper {
...
+            false -> accumulate(rex.type)
+        }
+    }
+
+    /**
+     * Checks for literal NULL or MISSING.
+     */
+    private fun Rex.isLiteralAbsent(): Boolean {
+        val op = this.op
+        return op is Rex.Op.Lit && (op.value.type == PartiQLValueType.MISSING || op.value.type == PartiQLValueType.NULL)
```
[the remark this answers is not in the document]
Comment (alancai98) [reply partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/DynamicTyper.kt:74]: followup to not call the function constructor for `missingValue` and `nullValue`,

```suggestion
        return op is Rex.Op.Lit && (op.value is NullValue || op.value is MissingValue)
```

File: partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt (hunk 8 above)
Comment (alancai98) [reply partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt:636]: I should have linked the comment for clarity -- https://github.com/partiql/partiql-lang-kotlin/pull/1463#discussion_r1604083403. We're calling `PartiQLValueType.MISSING` here but since it's deprecated in this PR and will be removed in a future release, I'm thinking we should not call the enum variant. Here's a version not calling the function invocation:

```suggestion
                op.value is MissingValue
```

File: partiql-types/src/main/kotlin/org/partiql/types/StaticType.kt [hunk 15]
```diff
@@ -170,6 +244,11 @@ public sealed class StaticType {
     /**
      * Type is optional if it is Any, or Missing, or an AnyOfType that contains Any or Missing type
      */
+    @Suppress("DEPRECATION")
+    @Deprecated(
+        message = "This will be removed in a future major-version bump.",
+        replaceWith = ReplaceWith("true")
+    )
```
[the remark this answers is not in the document]
Comment (johnedquinn) [reply partiql-types/src/main/kotlin/org/partiql/types/StaticType.kt:251]: True -- it was largely just to highlight the usages across the codebase.

File: partiql-types/src/main/kotlin/org/partiql/types/StaticType.kt [hunk 16]
```diff
@@ -621,7 +708,14 @@ public data class GraphType(
 /**
  * Represents a [StaticType] that's defined by the union of multiple [StaticType]s.
  */
-public data class AnyOfType(val types: Set<StaticType>, override val metas: Map<String, Any> = mapOf()) : StaticType() {
+public data class AnyOfType @Deprecated(
+    message = "This will be removed in a future major-version bump.",
```
[the remark this answers is not in the document]
Comment (johnedquinn) [reply partiql-types/src/main/kotlin/org/partiql/types/StaticType.kt:712]: Realistically, given our intention to eventually deprecate/delete this entire file, I don't really care to remove it. People should just not be using the constructor.

File: partiql-planner/src/test/kotlin/org/partiql/planner/internal/typer/PlanTyperTestsPorted.kt (hunk 12 above)
Comment (johnedquinn) [reply partiql-planner/src/test/kotlin/org/partiql/planner/internal/typer/PlanTyperTestsPorted.kt:2539]: Yeah, these tests originally are as follows:
```
--#[case-when-22]
-- type: (null|missing|int32)
CASE t_item.t_string
    WHEN 'a' THEN t_item.t_absent
    ELSE -1
END;

--#[case-when-23]
-- type: int32
-- false branch is pruned
CASE
    WHEN false THEN t_item.t_absent
    ELSE -1
END;
```

Since the t_absent type is now just ANY instead of `["null", "missing"]`, the result is just ANY -- and we have tests that already test this. That being said, I'm going to go one step further and delete the Ion tests and remove `t_absent` as well as `t_null` and `t_missing`.

File: partiql-planner/src/test/kotlin/org/partiql/planner/internal/typer/PlanTyperTestsPorted.kt (hunk 11 above)
Comment (johnedquinn) [reply partiql-planner/src/test/kotlin/org/partiql/planner/internal/typer/PlanTyperTestsPorted.kt:640]: Discussed offline -- but it has to do with the fact that a literal NULL has an unknown type, which gets interpreted as ANY for our type inference. Therefore, the applicable functions for `bitwise_and` are only for the integer types that correspond to the LHS being an int_32.

Before this PR, `1 & NULL` would be constant folded in the typer, which is wrong, because it doesn't actually give insight to the type being returned (just the value). This change narrows down the type. Though, in the future (or, if you'd like in this PR), we can attempt to find the highest precedence function when the argument is the literal NULL/MISSING value. This is kind of what PostgreSQL does. As it would result in more in-depth logic changes, I'd prefer a follow-up to this PR.

File: partiql-planner/src/test/kotlin/org/partiql/planner/internal/typer/PlanTyperTestsPorted.kt (hunk 1 above)
Comment (alancai98) [reply partiql-planner/src/test/kotlin/org/partiql/planner/internal/typer/PlanTyperTestsPorted.kt:2147]: Not too sure why the empty struct in the input binding tuples doesn't cause `ANY` for the type. Perhaps that binding tuple is filtered out?

Comment (alancai98) [reply partiql-planner/src/test/kotlin/org/partiql/planner/internal/typer/PlanTyperTestsPorted.kt:2147]: Not sure if there's a test showing this but if there's not, could still be helpful to include a test showing that the inferred type for a path expression on a literal struct with a `NULL`/`MISSING` value is `ANY`. E.g. `{ 'a': NULL }.a` -> `ANY` and `{ 'a': MISSING }.a` -> `ANY`

File: partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/TypeUtils.kt [hunk 17]
```diff
@@ -177,7 +195,7 @@ internal fun StructType.exclude(steps: List<Rel.Op.Exclude.Step>, lastStepOption
     val output = fields.map { field ->
         val newField = if (steps.size == 1) {
             if (lastStepOptional) {
-                StructType.Field(field.key, field.value.asOptional())
+                StructType.Field(field.key, field.value) // TODO: double check this
```
[the remark this answers is not in the document]
Comment (alancai98) [reply partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/TypeUtils.kt]: cool, thanks for cutting the issue. could also link the issue in the code for easier discoverability

**Review by RCHowell — CHANGES_REQUESTED** [review]

File: partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt [hunk 18]
```diff
@@ -301,14 +296,7 @@ internal class PlanTyper(
...
 
             // Calculate output schema given JOIN type
-            val l = lhs.type.schema
-            val r = rhs.type.schema
-            val schema = when (node.type) {
-                Rel.Op.Join.Type.INNER -> l + r
-                Rel.Op.Join.Type.LEFT -> l + r.pad()
-                Rel.Op.Join.Type.RIGHT -> l.pad() + r
-                Rel.Op.Join.Type.FULL -> l.pad() + r.pad()
-            }
```
[the remark this answers is not in the document]
Comment (johnedquinn) [reply partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt:311]: All types are considered nullable, therefore, there isn't a need to pad the types with the NULL type.

File: partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt [hunk 19]
```diff
@@ -515,20 +511,30 @@ internal class PlanTyper(
                     // we might improve upon this with some constant folding prior to typing
                     ANY
                 }
-            }.toSet()
-            val finalType = unionOf(pathTypes + toAddTypes).flatten()
-            return rex(finalType.swallowAny(), rexOpPathKey(root, key))
+            }
+            if (elementType.isEmpty()) {
+                handleAlwaysMissing()
+                return rex(ANY, rexOpPathKey(root, key))
```
[the remark this answers is not in the document]
Comment (johnedquinn) [reply partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt:517]: Yes, in the `v1` branch we can do that. Main doesn't have that concept yet.

File: partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt [hunk 20]
```diff
@@ -515,20 +511,30 @@ internal class PlanTyper(
...
 
         override fun visitRexOpPathSymbol(node: Rex.Op.Path.Symbol, ctx: StaticType?): Rex {
             val root = visitRex(node.root, node.root.type)
 
-            val paths = root.type.allTypes.map { type ->
-                val struct = type as? StructType ?: return@map rex(MISSING, rexOpLit(missingValue()))
+            // Check Root Type
+            if (!root.type.mayBeType<StructType>()) {
+                handleAlwaysMissing()
+                return rex(ANY, rexOpErr("Symbol lookup may only occur on structs, not ${root.type}."))
```
[the remark this answers is not in the document]
Comment (johnedquinn) [reply partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt:528]: https://github.com/partiql/partiql-lang-kotlin/pull/1463#discussion_r1610375092

File: partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt [hunk 21]
```diff
@@ -537,40 +543,28 @@ internal class PlanTyper(
                     )
                 }
             }
-            val type = unionOf(paths.map { it.type }.toSet()).flatten()
+            // Determine output type
+            val type = when (paths.size) {
+                // Escape early since no inference could be made
+                0 -> {
+                    handleAlwaysMissing()
+                    return rex(ANY, Rex.Op.Path.Symbol(root, node.key))
```
[the remark this answers is not in the document]
Comment (johnedquinn) [reply partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt:551]: https://github.com/partiql/partiql-lang-kotlin/pull/1463#discussion_r1610375092

**Review by RCHowell — APPROVED** [review]
